File: reference/typescript/src/tool/amcp.ts
import { SPEC_VERSION, type AuditEvent } from '../schema/event.js';
import { hashCanonical } from '../ledger/canonical.js';
import { computeRecordHash, countersignaturePayload } from '../ledger/ledger.js';
import type { AttemptResponse, AuditTransport } from '../transport/transport.js';
import type { EventSigner } from '../l2/signing.js';
import { AuditAttemptResultSchema } from '../transport/mcpWire.js';

type AbortReason = NonNullable<AuditEvent['reason']>;
type EventOf = (outcome: AuditEvent['outcome'], reason?: AbortReason) => AuditEvent;

// Tool-side audit-before-act (§6): emit attempt, await durable accept, perform the domain
// action, emit outcome. reject or unavailable => action not performed.

// Effect axis (§4.2). Both flags declared per operation; action_type is opaque, so no inference.
export interface Effect {
  mutates: boolean;
  egress: boolean;
}

// Tool-side fail-closed halt: the tool did not perform the domain action (outcome=aborted). Named
// for the tool's own abort, not host "blocking" - the host never prevents a domain action (§2).
export class AmcpAbortedError extends Error {
  constructor(
    readonly action_type: string,
    readonly target_ref: string,
    readonly reason: string,
  ) {
    super(`auditable-mcp aborted ${action_type} on ${target_ref}: ${reason}`);
    this.name = 'AmcpAbortedError';
  }
}

export interface ActionSpec {
  action_type: string;
  target_resource: AuditEvent['target_resource'];
  effect: Effect;
  // Confidentiality is the tool's choice (§4.3): disclose cleartext, seal a hash, both, or neither.
  disclose?: Record<string, unknown>; // -> event.action_context
  commit?: unknown; // -> event.action_context_hash = sha256(canonical(commit))
}

export interface AmcpDeps {
  newId: () => string;
  now: () => string; // ISO datetime
}

// Resolves a `host_key_id` and verifies a countersignature over the host-assigned fields and log_id (§7.2).
export type CountersignatureVerifier = (hostKeyId: string, signature: string, payload: string) => boolean;

export class AmcpSession {
  constructor(
    private readonly transport: AuditTransport,
    // The audit session the host issued for this call (§6.3), carried verbatim in every event.
    private readonly sessionId: string,
    private readonly deps: AmcpDeps,
    // Signer presence is the only L1/L2 emission difference (§5).
    private readonly signer?: EventSigner,
    // A tool that requires `countersign: "host"` (§5.2) verifies every accept and aborts rather than
    // act on a record no distinct party confirmed (§7.2). Requiring it without a verifier would
    // abort every action against a host that is signing correctly, so the pairing is enforced.
    private readonly countersignatureVerifier?: CountersignatureVerifier,
    private readonly requireCountersign = false,
    // How many times the identical attempt is sent again after `unavailable` or no answer (§7.1).
    private readonly attemptRetries = 0,
  ) {
    if (requireCountersign && countersignatureVerifier === undefined) {
      throw new Error('requireCountersign needs a CountersignatureVerifier (§7.2, §11.3)');
    }
  }

  private stamp(event: AuditEvent): AuditEvent {
    return this.signer ? this.signer.sign(event) : event;
  }

  // One Attempt Response, or undefined when the attempt went unanswered: a transport fault, a
  // protocol error, or an answer that does not validate against the Attempt Response schema - a
  // partial countersignature triple among them - is not an Attempt Response (§6).
  private async ask(attempt: AuditEvent): Promise<AttemptResponse | undefined> {
    let answer: unknown;
    try {
      answer = await this.transport.sendAttempt(attempt);
    } catch {
      return undefined;
    }
    const parsed = AuditAttemptResultSchema.safeParse(answer);
    return parsed.success ? (answer as AttemptResponse) : undefined;
  }

  private async abort(event: EventOf, spec: ActionSpec, reason: AbortReason): Promise<never> {
    await this.transport.sendOutcome(this.stamp(event('aborted', reason)));
    throw new AmcpAbortedError(spec.action_type, spec.target_resource.ref, reason);
  }

  // Wrap one domain operation in audit-before-act (§6). `perform` runs at most once, and only after an
  // accept that passed every check of §7.2 (§6).
  async audited<T>(spec: ActionSpec, perform: () => Promise<T>): Promise<T> {
    const head = {
      id: this.deps.newId(),
      spec_version: SPEC_VERSION,
      ts: this.deps.now(),
      session_id: this.sessionId,
      action_type: spec.action_type,
      mutates: spec.effect.mutates,
      egress: spec.effect.egress,
      target_resource: spec.target_resource,
    } as const;
    const context = {
      ...(spec.disclose !== undefined ? { action_context: spec.disclose } : {}),
      ...(spec.commit !== undefined ? { action_context_hash: hashCanonical(spec.commit) } : {}),
    };
    // Members in the order the schema lists them.
    const event: EventOf = (outcome, reason) => ({ ...head, outcome, ...(reason !== undefined ? { reason } : {}), ...context });

    // Signed once: a retry sends the byte-identical attempt, which keeps its signer_seq (§6, §7.1).
    const attempt = this.stamp(event('attempted'));
    let resp = await this.ask(attempt);
    for (let retry = 0; retry < this.attemptRetries && (resp === undefined || resp.status === 'unavailable'); retry += 1) {
      resp = await this.ask(attempt);
    }
    if (resp === undefined || resp.status !== 'accept') {
      // reject, unavailable, or unanswered: do not act; signal aborted (§7.2).
      return this.abort(event, spec, resp?.status === 'reject' ? 'host-rejected' : 'host-unavailable');
    }

    // §7.2 evaluates in precedence order: the response's status above, then the countersignature
    // that authenticates the host-assigned fields, then the hash computed over them. The reason is
    // sealed into the ledger and compared across implementations, so the order is not incidental.
    if (this.requireCountersign && resp.host_signature === undefined) {
      return this.abort(event, spec, 'host-uncountersigned');
    }
    if (resp.host_signature !== undefined && this.countersignatureVerifier !== undefined) {
      const payload = countersignaturePayload(resp.seq, resp.host_ts, resp.log_id as string, resp.previous_hash, resp.record_hash);
      if (!this.countersignatureVerifier(resp.host_key_id as string, resp.host_signature, payload)) {
        return this.abort(event, spec, 'host-signature-invalid');
      }
    }

    // Polluted Stop (§7.2): MUST under L2 (signer present) and wherever a countersignature is
    // required, OPTIONAL otherwise. The countersignature binds the host-assigned fields to
    // record_hash and no further; recomputing record_hash over the attempt bytes is what binds it
    // to this attempt, so an accept for another record is refused.
    if (this.signer !== undefined || this.requireCountersign) {
      const expected = computeRecordHash(attempt, resp.seq, resp.host_ts, resp.previous_hash);
      if (expected !== resp.record_hash) return this.abort(event, spec, 'hash-mismatch');
    }

    let result: T;
    try {
      result = await perform();
    } catch (err) {
      await this.emitOutcome(event('failed'));
      throw err;
    }
    // Outside the `try`: an action that was performed is never recorded `failed`, and a lost outcome is a
    // completeness gap the host resolves (§10.8), not the caller's error to retry.
    await this.emitOutcome(event('success'));
    return result;
  }

  /** Send a terminal outcome, logging rather than throwing if it cannot be delivered (§6, §10.8). */
  private async emitOutcome(outcome: AuditEvent): Promise<void> {
    try {
      await this.transport.sendOutcome(this.stamp(outcome));
    } catch (err) {
      console.error(`could not deliver the ${outcome.outcome} outcome of ${outcome.id}; it is lost`, err);
    }
  }
}

// Deterministic deps for reproducible test vectors: no wall clock, no random uuid.
export function deterministicDeps(): AmcpDeps {
  let n = 0;
  return {
    newId: () => {
      n += 1;
      // uuid-shaped id to satisfy the schema uuid() check.
      const h = n.toString(16).padStart(12, '0');
      return `00000000-0000-4000-8000-${h}`;
    },
    now: () => {
      // Fixed base + counter; stable ts across runs.
      const secs = 1000 + n;
      return new Date(secs * 1000).toISOString();
    },
  };
}

File: reference/python/src/auditable_mcp/amcp.py
"""Tool-side Auditable MCP session.

Enforces the audit-before-act discipline:
1. Emit attempt event.
2. Await durable accept from host.
3. Perform internal domain action (not performed if rejected/unavailable).
4. Emit outcome event (success/failed).
"""

import logging
from collections.abc import Callable
from datetime import UTC, datetime
from typing import Protocol, TypeVar

from auditable_mcp.canonical import hash_canonical
from auditable_mcp.ledger import compute_record_hash, countersignature_payload
from auditable_mcp.schema import validate_attempt_response
from auditable_mcp.transport import AttemptResponse, AuditTransport, AuditTransportError, response_members

logger = logging.getLogger(__name__)

SPEC_VERSION = 'auditable-mcp/0.3'
_BASE_EPOCH_SECONDS = 1000

T = TypeVar('T')


class EventSigner(Protocol):
    """Signs an event, stamping key_id, signer_seq, and signature (Level 2)."""

    def sign(self, event: dict) -> dict:
        """Return the signed event."""
        ...


class AmcpAbortedError(Exception):
    """Tool-side fail-closed halt: the domain action was not performed (outcome=aborted).

    Named for the tool's own abort, not host "blocking" - the host never prevents a domain
    action (§2).
    """

    def __init__(self, action_type: str, target_ref: str, reason: str) -> None:
        """Capture the aborted action for the tools/call result."""
        super().__init__(f'auditable-mcp aborted {action_type} on {target_ref}: {reason}')
        self.action_type = action_type
        self.target_ref = target_ref
        self.reason = reason


class DeterministicDeps:
    """Deterministic id and timestamp source for reproducible runs (no wall clock / uuid4)."""

    def __init__(self) -> None:
        """Start the counter at zero."""
        self._n = 0

    def new_id(self) -> str:
        """Return the next deterministic uuid-shaped id."""
        self._n += 1
        return f'00000000-0000-4000-8000-{self._n:012x}'

    def now(self) -> str:
        """Return a deterministic ISO-8601 timestamp string."""
        moment = datetime.fromtimestamp(_BASE_EPOCH_SECONDS + self._n, tz=UTC)
        return moment.strftime('%Y-%m-%dT%H:%M:%S.000Z')


class CountersignatureVerifier(Protocol):
    """Resolves a `host_key_id` and verifies a countersignature over the host-assigned fields and log_id."""

    def __call__(self, host_key_id: str, signature: str, payload: str) -> bool:
        """Return True only if the signature verifies against the key the registry binds (§7.2)."""
        ...


class AmcpSession:
    """Wraps internal operations in the audit-before-act lifecycle.

    L1 and L2 emission are identical; L2 only adds a signer.
    """

    def __init__(
        self,
        transport: AuditTransport,
        session_id: str,
        deps: DeterministicDeps,
        signer: EventSigner | None = None,
        countersignature_verifier: CountersignatureVerifier | None = None,
        require_countersign: bool = False,
        attempt_retries: int = 0,
    ) -> None:
        """Bind the session to a transport, the call's audit session, id/time deps, and optional signer.

        `session_id` is the audit session the host issued for this call (§6.3), carried verbatim in
        every event. A tool that declares `countersign: "host"` (§5.2) requires every accept to carry
        a countersignature it can verify, so `require_countersign` needs a verifier: without one the
        tool would abort every action on a host that is countersigning correctly. `attempt_retries` is
        how many times the identical attempt is sent again after `unavailable` or no answer (§7.1).

        Raises:
            ValueError: `require_countersign` was set without a verifier.
        """
        if require_countersign and countersignature_verifier is None:
            raise ValueError('require_countersign needs a CountersignatureVerifier (§7.2, §11.3)')
        self._transport = transport
        self._session_id = session_id
        self._deps = deps
        self._signer = signer
        self._countersignature_verifier = countersignature_verifier
        self._require_countersign = require_countersign
        self._attempt_retries = attempt_retries

    def _stamp(self, event: dict) -> dict:
        """Sign the event if a signer is present (L2), else return it unchanged (L1)."""
        return self._signer.sign(event) if self._signer is not None else event

    def audited(
        self,
        action_type: str,
        target_resource: dict,
        perform: Callable[[], T],
        *,
        mutates: bool,
        egress: bool,
        disclose: dict | None = None,
        commit: object | None = None,
    ) -> T:
        """Emit attempt, await accept, perform the action, then emit the outcome.

        The effect axis (mutates, egress) is declared explicitly per operation. Confidentiality
        is the tool's choice (§4.3): ``disclose`` records cleartext params, ``commit`` records
        a hash of the exact input; either, both, or neither may be given.
        """
        base = {
            'id': self._deps.new_id(),
            'spec_version': SPEC_VERSION,
            'ts': self._deps.now(),
            'session_id': self._session_id,
            'action_type': action_type,
            'mutates': mutates,
            'egress': egress,
            'target_resource': target_resource,
        }
        if disclose is not None:
            base['action_context'] = disclose
        if commit is not None:
            base['action_context_hash'] = hash_canonical(commit)
        # Signed once: a retry sends the byte-identical attempt, which keeps its signer_seq (§6, §7.1).
        attempt = self._stamp({**base, 'outcome': 'attempted'})
        response = self._ask(attempt)
        for _ in range(self._attempt_retries):
            if response is not None and response.status != 'unavailable':
                break
            response = self._ask(attempt)
        if response is None or response.status != 'accept':
            # reject, unavailable, or unanswered: do not act; signal aborted (§7.2).
            reason = 'host-rejected' if response is not None and response.status == 'reject' else 'host-unavailable'
            self._abort(base, attempt['id'], reason)
            raise AmcpAbortedError(action_type, target_resource['ref'], reason)
        # §7.2 evaluates in precedence order: the response's status above, then the countersignature
        # that authenticates the host-assigned fields, then the hash computed over them.
        # The reason is sealed into the ledger and compared across implementations, so the order is
        # not incidental.
        if self._require_countersign and response.host_signature is None:
            self._abort(base, attempt['id'], 'host-uncountersigned')
            raise AmcpAbortedError(action_type, target_resource['ref'], 'host-uncountersigned')
        if response.host_signature is not None and self._countersignature_verifier is not None:
            payload = countersignature_payload(
                response.seq, response.host_ts, response.log_id, response.previous_hash, response.record_hash
            )
            if not self._countersignature_verifier(response.host_key_id, response.host_signature, payload):
                self._abort(base, attempt['id'], 'host-signature-invalid')
                raise AmcpAbortedError(action_type, target_resource['ref'], 'host-signature-invalid')
        # Polluted Stop (§7.2): MUST under L2 (signer present) and wherever a countersignature is
        # required, OPTIONAL otherwise. The countersignature binds the host-assigned fields to
        # record_hash and no further; recomputing record_hash over the attempt bytes is what binds it
        # to this attempt, so an accept for another record is refused.
        if self._signer is not None or self._require_countersign:
            expected = compute_record_hash(attempt, response.seq, response.host_ts, response.previous_hash)
            if expected != response.record_hash:
                self._abort(base, attempt['id'], 'hash-mismatch')
                raise AmcpAbortedError(action_type, target_resource['ref'], 'hash-mismatch')
        # `perform` runs at most once, and only after an accept that passed every check above (§6).
        try:
            result = perform()
        except Exception:
            self._emit_outcome({**base, 'id': attempt['id'], 'outcome': 'failed'})
            raise
        # Outside the `try`: an action that was performed is never recorded `failed`, and a lost outcome
        # is a completeness gap the host resolves (§10.8), not the caller's error to retry.
        self._emit_outcome({**base, 'id': attempt['id'], 'outcome': 'success'})
        return result

    def _emit_outcome(self, event: dict) -> None:
        """Send a terminal outcome, logging rather than raising if it cannot be delivered (§6, §10.8)."""
        try:
            self._transport.send_outcome(self._stamp(event))
        except Exception:
            logger.exception('could not deliver the %s outcome of %s; it is lost', event['outcome'], event['id'])

    def _ask(self, attempt: dict) -> AttemptResponse | None:
        """Send the attempt and return the Attempt Response, or None when it went unanswered.

        A transport fault, a protocol error, or an answer that does not validate against the Attempt
        Response schema - a partial countersignature triple among them - is not an Attempt Response,
        and the tool treats it exactly as `unavailable` (§6).
        """
        try:
            response = self._transport.send_attempt(attempt)
        except (AuditTransportError, OSError, TimeoutError):
            return None
        if (
            not isinstance(response, AttemptResponse)
            or validate_attempt_response(response_members(response)) is not None
        ):
            return None
        return response

    def _abort(self, base: dict, event_id: str, reason: str) -> None:
        """Emit an aborted outcome recording why the domain action was not performed."""
        self._transport.send_outcome(self._stamp({**base, 'id': event_id, 'outcome': 'aborted', 'reason': reason}))
